app/models/metric/processing.rb

```diff
@@ -178,4 +118,108 @@ def self.interval_name_to_interval(name)
     else raise _("unknown interval name: [%{name}]") % {:name => name}
     end
   end
+
+  ##### DEFINING METHODS FOR .process_derived_columns #####
+
+  # Shared_methods
+  def self.cpu_metrics?(attrs)
+    attrs[:cpu_usage_rate_average] || attrs[:cpu_usagemhz_rate_average]
+  end
+  private_class_method :cpu_metrics?
+
+  def self.mem_metrics?(attrs)
+    attrs[:mem_usage_absolute_average] || attrs[:derived_memory_used]
+  end
+  private_class_method :mem_metrics?
+
+  # Defines:
+  #   total_cpu                         (helper method)
+  #   total_mem                         (helper method)
+  #   generate_derived_cpu_available
+  #   generate_derived_cpu_reserved     (calls state.reserve_cpu)
+  #   generate_derived_memory_available
+  #   generate_derived_memory_reserved  (calls state.reserve_mem)
+  %w[cpu memory].each do |group|
+    method_def = <<-METHOD_DEF
+      def self.total_#{group[0..2]}(state)
+        state.total_#{group[0..2]} || 0
+      end
+      private_class_method :total_#{group[0..2]}
+
+      def self.generate_derived_#{group}_available(col, state, attrs, result)
+        result[col] = total_#{group[0..2]}(state) if #{group[0..2]}_metrics?(attrs) && total_#{group[0..2]}(state) > 0
+      end
+      private_class_method :generate_derived_#{group}_available
+
+      def self.generate_derived_#{group}_reserved(col, state, _, result)
+        result[col] = state.reserve_#{group[0..2]}
+      end
+      private_class_method :generate_derived_#{group}_reserved
+    METHOD_DEF
+    eval method_def.split("\n").map(&:strip).join(';')
```

Comment:
Assuming we went with the `eval`, why the need for all of the split/map/join?  I thought eval just worked with the original full string...

```
[1] pry(main)> class Foo
[1] pry(main)*   %w{a b}.each do |x|
[1] pry(main)*     eval <<-EOS
[1] pry(main)* def #{x}
[1] pry(main)*   puts "#{x}"
[1] pry(main)* end
[1] pry(main)*     EOS
[1] pry(main)*   end
[1] pry(main)* end
=> ["a", "b"]
[2] pry(main)> Foo.new.a
a
=> nil
[3] pry(main)> Foo.new.b
b
=> nil
```

Reply:
Yeah, probably could go without it, but I think I was trying to channel what was done here to a degree (probably not needed though):

https://github.com/rails/rails/commit/44c51fc (see associated issue as well).

I would have to check to see if that even helps though.  Honestly, I forget since it has been a while since I have put this code together.  Will test out in a bit.

Reply:
The Rails commit link there doesn't work.

Reply:
@Fryguy Shoot... sorry, I usually self shorten the SHA's to 7 chars, and I guess that was the one time Rails actually had a conflicting commit SHA with that... updated in that last comment so there isn't a broken link.

Regardless, it was stupid to do that in this case since I wasn't building a DSL, so that probably shouldn't stay.  Line numbers would be exactly where we want them in this case.